internal/servers/controller/listeners.go

```diff
@@ -168,6 +83,104 @@ func (c *Controller) startListeners(ctx context.Context) error {
+
+	handler, err := c.apiHandler(HandlerProperties{
+		ListenerConfig: ln.Config,
+		CancelCtx:      ctx,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	cancelCtx := c.baseContext // Resolve to avoid race conditions if the base context is replaced.
+	server := &http.Server{
+		Handler:           handler,
+		ReadHeaderTimeout: 10 * time.Second,
+		ReadTimeout:       30 * time.Second,
+		IdleTimeout:       5 * time.Minute,
+		ErrorLog:          c.logger.StandardLogger(nil),
+		BaseContext:       func(net.Listener) context.Context { return cancelCtx },
+	}
+	ln.HTTPServer = server
+
+	if ln.Config.HTTPReadHeaderTimeout > 0 {
+		server.ReadHeaderTimeout = ln.Config.HTTPReadHeaderTimeout
+	}
+	if ln.Config.HTTPReadTimeout > 0 {
+		server.ReadTimeout = ln.Config.HTTPReadTimeout
+	}
+	if ln.Config.HTTPWriteTimeout > 0 {
+		server.WriteTimeout = ln.Config.HTTPWriteTimeout
+	}
+	if ln.Config.HTTPIdleTimeout > 0 {
+		server.IdleTimeout = ln.Config.HTTPIdleTimeout
+	}
+
+	switch ln.Config.TLSDisable {
+	case true:
+		l, err := ln.Mux.RegisterProto(alpnmux.NoProto, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error getting non-tls listener: %w", err)
+		}
+		if l == nil {
+			return nil, errors.New("could not get non-tls listener")
+		}
+		apiServers = append(apiServers, func() { go server.Serve(l) })
+
+	default:
+		protos := []string{"", "http/1.1", "h2"}
+		for _, v := range protos {
+			l := ln.Mux.GetListener(v)
+			if l == nil {
+				return nil, fmt.Errorf("could not get tls proto %q listener", v)
+			}
+			apiServers = append(apiServers, func() { go server.Serve(l) })
+		}
+	}
+
+	return apiServers, nil
+}
+
+func (c *Controller) configureForCluster(ctx context.Context, ln *base.ServerListener) error {
+	// Clear out in case this is a second start of the controller
+	ln.Mux.UnregisterProto(alpnmux.DefaultProto)
+	l, err := ln.Mux.RegisterProto(alpnmux.DefaultProto, &tls.Config{
+		GetConfigForClient: c.validateWorkerTls,
+	})
+	if err != nil {
+		return fmt.Errorf("error getting sub-listener for worker proto: %w", err)
+	}
+
+	workerReqInterceptor, err := workerRequestInfoInterceptor(ctx, c.conf.Eventer)
+	if err != nil {
+		return fmt.Errorf("error getting sub-listener for worker proto: %w", err)
+	}
+
+	workerServer := grpc.NewServer(
+		grpc.MaxRecvMsgSize(math.MaxInt32),
+		grpc.MaxSendMsgSize(math.MaxInt32),
+		grpc.UnaryInterceptor(
+			grpc_middleware.ChainUnaryServer(
+				workerReqInterceptor,
+				auditRequestInterceptor(ctx),  // before we get started, audit the request
+				auditResponseInterceptor(ctx), // as we finish, audit the response
+			),
+		),
+	)
+
+	workerService := workers.NewWorkerServiceServer(c.ServersRepoFn, c.SessionRepoFn, c.workerStatusUpdateTimes, c.kms)
+	pbs.RegisterServerCoordinationServiceServer(workerServer, workerService)
+	pbs.RegisterSessionServiceServer(workerServer, workerService)
+
+	interceptor := newInterceptingListener(c, l)
+	ln.ALPNListener = interceptor
+	ln.GrpcServer = workerServer
```

Comment:
I've found these listeners to be a great place to store this state. They provide some key benefits:
- Means we don't have to return something like `[]func()` in this func;
- We can easily correlate the listener config and the server/listener that was created and ultimately will be started to serve it. This is important to support our future efforts of starting these listeners more intelligently.
- Conceptually speaking, they seem like the correct place to be storing this state in.
- If everything is stored centrally as it would be this way, shutting the servers down becomes easy.

But there's some problems with this that need to be addressed if we're going this way. More details [here](https://github.com/hashicorp/boundary/pull/1874/files#r811193116)

Reply:
Considering we only need access to the `Serve` functions ephemerally, I decided to go with returning functions.